### PR TITLE
expose ToElement method from ElementModelProvider

### LINKED ISSUE #172: Using strong-typed element in Language Variants
### Motivation

**_Why is this feature required? What problems does it solve?_**
Being able to retrieve a strong-typed language variant, and/or its elements when processing content items

### Proposed solution

Add an extension method to `LanguageVariantModel` that makes use of the functionality in `Kentico.Kontent.Management.Modules.ModelBuilders.ElementModelProvider` to return all of the boxed element in an `IEnumerable<BaseElement>`. Also, possibly exposing the `ToElement(...)`.

### Additional context

The reason I came across this issue is that I'm writing an integration with a 3<sup>rd</sup> party translation service. The service uses a mixture of machine and human translation, so providing to a readable structure is a must. T do this I _currently_ use the Devliery API (DAPI) rather than the Management API (MAPI) and recusrively create a JSON document that presents a hierarchy of a content item and it 'children', focussing on Text, RichText element. This looks similar to the below:
```json
{
  "ItemType": "faq_section",
  "ItemCodename": "test_faq_5c6875b",
  "ItemElements": [
    {
      "ElementCodename": "title",
      "ElementType": "text",
      "ElementValue": "This is a test FAQ"
    },
    {
      "ElementCodename": "questions",
      "ElementType": "modular_content",
      "ElementItems": [
        {
          "ItemType": "faq",
          "ItemCodename": "test_faq__03",
          "ItemElements": [
            {
              "ElementCodename": "answer",
              "ElementType": "rich_text",
              "ElementValue": "\u003Cp\u003E42\u003C/p\u003E"
            },
            {
              "ElementCodename": "question",
              "ElementType": "rich_text",
              "ElementValue": "\u003Cp\u003EWhat is the meaning of life?\u003C/p\u003E"
            }
          ]
        }
      ]
    }
  ]
}
```

`RichText` however obviously provides some complications due to modular content and component etc. These cannot be resolved for translation, so I want to use the MAPI to deal with exporting the `RichText` fields. Possibly if things get easier I can also use MAPI for the entire export, though the rate limit on the MAPI may be prohibitive. Due to how this is working, I don't know the Conent Item's type at design time and dependon on what can be determined at runtime.

I'm using the following code to work things through:
```csharp
var mapiLanguageVariant = await mapiClient.GetLanguageVariantAsync(
  new LanguageVariantIdentifier(
    Reference.ByCodename(ITEM_CODENAME),
    Reference.ByCodename(LANGUAGE_CODENAME)
  ));
var mapiContentItem = await mapiClient.GetContentItemAsync(
  mapiLanguageVariant.Item
  );
var mapiContentType = await mapiClient.GetContentTypeAsync(
  mapiContentItem.Type
  );


foreach (var mapiTypeElement in mapiContentType.Elements)
{
  switch (mapiTypeElement.Type)
  {
    case ElementMetadataType.RichText:
      // TODO : Something amazing
      var mapiItemElement = (dynamic) mapiLanguageVariant.Elements.FirstOrDefault(element => Guid.Parse(element.element.id) == mapiTypeElement.Id);
      break;
  }
}
```
As it stands, `mapiItemElement` is an `ExpandoObject`, so needs to be cast using `(dynamic)mapiItemElement` before the `components`, `element`, and `value` properties can be accessed. Copying the `ToElement` method as mentioned above provides a much smoother and more readable codebase:
```csharp
var typedElement = CustomElementModelProvider.ToElement((dynamic) mapiItemElement, typeof(RichTextElement));
```
We now have `typedElement.Components` 👌 However, being able to do that with the `mapiLanguageVariant` as an extension method would make this a simpler task.  I.e. 
```csharp
IEnumerable<BaseElement> typedElements = mapiLanguageVariant.GetTypedElements(mapiContentType.Elements);
``` 
(if we can do it without passing the type in, that's even better!)

### PR DESCRIPTION
**Pull request by winklertomas**
### Motivation

Which issue does this fix? Fixes #172 

### Checklist

- [ ] Code follows coding conventions held in this repo
- [ ] Tests are passing

### CODE REVIEW

File: Kentico.Kontent.Management/Models/LanguageVariants/Elements/BaseElement.cs [hunk 1]
```diff
@@ -20,6 +23,136 @@ public abstract class BaseElement
...
+    /// <summary>
+    /// Transforms the dynamic object to the element.
+    /// </summary>
+    public static BaseElement FromDynamic(dynamic source, Type type)
+    {
+        if (type == typeof(TextElement))
+        {
+            return new TextElement
+            {
+                // TODO extend by codename + external ID
```
Comment (IvanKiral) [inline Kentico.Kontent.Management/Models/LanguageVariants/Elements/BaseElement.cs:54]: After consulting with Ondra, the id does not need to be set, in that case the program should atlest throw some exception, or the other option is to add another option (e.g byDynamic) which finds out which of the reference variable is set and then use it.

Comment (gormal) [reply Kentico.Kontent.Management/Models/LanguageVariants/Elements/BaseElement.cs:54]: This is probably copy-pasted from ElementModelProvider, there, this comment makes sense as API always returns only Id. Now as we make it public to end-users the source might contain anything. We need to handle validation more robustly (e.g. what if the source object has a totally different shape than the type expects?). 


When it comes to the comment one possible approach is to try to find Id then Codename and lastly ExternalId and use the first to be found.

@winklertomas could you please have a look at the possible validation of inputs?

PS: The solution "it is good enough right now, can be improved later" is also valid.

**Review by gormal — CHANGES_REQUESTED** [review]
The method `FromDynamic` is publicly accessible now, so we should consider input validation.

File: Kentico.Kontent.Management/Extensions/DynamicExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,18 @@
+﻿using System.Collections.Generic;
+using System.Dynamic;
+
+namespace Kentico.Kontent.Management.Extensions
+{
+    public static class DynamicExtensions
```
Comment (gormal) [inline Kentico.Kontent.Management/Extensions/DynamicExtensions.cs:6]: Please add an xml comment.

File: Kentico.Kontent.Management/Extensions/DynamicExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    {
+        public static bool HasProperty(dynamic source, string propertyName)
```
Comment (gormal) [inline Kentico.Kontent.Management/Extensions/DynamicExtensions.cs:8]: Please add an xml comment.

File: Kentico.Kontent.Management.Tests/Modules/Extensions/DynamicExtensionsTests.cs [hunk 4]
```diff
@@ -0,0 +1,46 @@
+﻿using Kentico.Kontent.Management.Extensions;
+using Kentico.Kontent.Management.Modules.Extensions;
+using Kentico.Kontent.Management.Modules.ModelBuilders;
+using System;
```
Comment (gormal) [inline Kentico.Kontent.Management.Tests/Modules/Extensions/DynamicExtensionsTests.cs:4]: Please, remove unused usings.

File: Kentico.Kontent.Management/Extensions/DynamicExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,18 @@
+﻿using System.Collections.Generic;
+using System.Dynamic;
+
+namespace Kentico.Kontent.Management.Extensions
```
Comment (gormal) [inline Kentico.Kontent.Management/Extensions/DynamicExtensions.cs:4]: Convert to file-scoped namespace, please.

File: Kentico.Kontent.Management/Models/LanguageVariants/Elements/BaseElement.cs [hunk 6]
```diff
@@ -21,34 +25,140 @@ public abstract class BaseElement
...
+        ElementMetadataType.Subpages => FromDynamic(source, typeof(SubpagesElement)),
+        _ => throw new InvalidOperationException("Unknown element type"),
+    };
+
+    /// <summary>
+    /// Transforms the dynamic object to the element.
+    /// </summary>
+    public static BaseElement FromDynamic(dynamic source, Type type)
+    {
+        ValidateSource(source);
```
Comment (gormal) [inline Kentico.Kontent.Management/Models/LanguageVariants/Elements/BaseElement.cs:51]: Even though there is validation in place, dynamic still can have an unexpected shape, or the value property might have an unexpected type. I suggest that we put the whole FromDynamic into try-catch block and throw an exception with a message like "Object could not be converted to the strongly-typed element. Please check if it has expected properties with expected type".

File: Kentico.Kontent.Management/Models/LanguageVariants/Elements/BaseElement.cs [hunk 7]
```diff
@@ -21,34 +25,140 @@ public abstract class BaseElement
...
             };
         }
 
-        throw new DataMisalignedException("Element reference does not contain any identifier.");
+        throw new ArgumentOutOfRangeException($"{type} is not a valid element");
+    }
+
+    private static void ValidateSource(dynamic source)
+    {
+        if (source == null || !DynamicExtensions.HasProperty(source, "element") || !DynamicExtensions.HasProperty(source, "value"))
```
Comment (gormal) [inline Kentico.Kontent.Management/Models/LanguageVariants/Elements/BaseElement.cs:159]: Additional check could be `source.GetType().BaseType == typeof(BaseElement)`

File: Kentico.Kontent.Management/Models/Shared/Reference.cs [hunk 8]
```diff
@@ -45,4 +46,58 @@ public sealed class Reference
     /// </summary>
     /// <param name="externalId">The external id of the identifier.</param>
     public static Reference ByExternalId(string externalId) => new() { ExternalId = externalId };
+
+    /// <summary>
+    /// Transforms the dynamic object to the <see cref="Reference"/>
+    /// </summary>
+    public static Reference FromDynamic(dynamic source)
+    {
+        if (DynamicExtensions.HasProperty(source, "id"))
```
Comment (gormal) [inline Kentico.Kontent.Management/Models/Shared/Reference.cs:55]: Again, we should be more defensive and put this whole FromDynamic into try-catch block and throw a descriptive exception if the Reference object cannot be created.

**Review by gormal — CHANGES_REQUESTED** [review]
Great job, I have only a few nitpicks and one concern - code should be more defensive when converting from a dynamic object to a strongly-typed object.

File: Kentico.Kontent.Management/Models/LanguageVariants/Elements/BaseElement.cs (hunk 7 above)
Comment (winklertomas) [reply Kentico.Kontent.Management/Models/LanguageVariants/Elements/BaseElement.cs:159]: In my opinion, the basetype doesn't have to be the BaseElement, it is enough to have the required properties

**Review by gormal — APPROVED** [review]

**Comment by Simply007** [comment]
We should cover this functionality in README.

**Pull request merged**